src/rust/engine/process_executor/src/test_server.rs

```diff
@@ -0,0 +1,159 @@
+}
+
+impl TestServer {
+  ///
+  /// # Arguments
+  /// * `mock_execution` - The canned responses to issue. Returns the MockExecution's
+  ///                      operation_responses in order to any ExecuteRequest or GetOperation
+  ///                      requests.
+  ///
+  ///                      If an ExecuteRequest request is received which is not equal to this
+  ///                      MockExecution's execute_request, an error will be returned.
+  ///
+  ///                      If a GetOperation request is received whose name is not equal to this
+  ///                      MockExecution's name, or more requests are received than stub responses
+  ///                      are available for, an error will be returned.
+  pub fn new(mock_execution: MockExecution) -> TestServer {
+    let mock_server = MockServer::new(mock_execution);
+
+    let env = Arc::new(grpcio::Environment::new(1));
+    let mut server_transport = grpcio::ServerBuilder::new(env)
+        .register_service(bazel_protos::remote_execution_grpc::create_execution(mock_server.clone()))
+        .register_service(bazel_protos::operations_grpc::create_operations(mock_server.clone()))
+        .bind("localhost", 0)
+        .build()
+        .unwrap();
+    server_transport.start();
+
+    TestServer {
+      mock_server,
+      server_transport,
+    }
+  }
+
+  ///
+  /// The address on which this server is listening over insecure HTTP transport.
+  ///
+  pub fn address(&self) -> String {
+    let bind_addr = self.server_transport.bind_addrs().first().unwrap();
+    format!("{}:{}", bind_addr.0, bind_addr.1)
+  }
+}
+
+impl Drop for TestServer {
+  fn drop(&mut self) {
+    let expected_responses = self.mock_server.mock_execution.operation_responses.len();
+    let issued_responses = self.mock_server.next_operation_response.lock().unwrap().deref().clone();
+    assert_eq!(
+      expected_responses,
+      issued_responses,
+      "Expected {} responses but only issued {}. Expected responses:\n{}\nReceived requests:\n{}",
+      expected_responses,
+      issued_responses,
+      MockServer::display_all(self.mock_server.mock_execution.operation_responses.iter()),
+      MockServer::display_all(self.mock_server.received_messages.deref().lock().unwrap().iter())
+    )
+  }
+}
+
+#[derive(Clone, Debug)]
+pub struct MockServer {
+  mock_execution: MockExecution,
+  next_operation_response: Arc<Mutex<usize>>,
+  received_messages: Arc<Mutex<Vec<(String, Box<protobuf::Message>)>>>,
+}
+
+impl MockServer {
+  fn new(mock_execution: MockExecution) -> MockServer {
+    MockServer {
+      mock_execution: mock_execution,
+      next_operation_response: Arc::new(Mutex::new(0)),
+      received_messages: Arc::new(Mutex::new(vec![])),
+    }
+  }
+
+  fn log<T: protobuf::Message + Sized>(&self, message: T) {
+    self.received_messages.lock().unwrap().push((message.descriptor().name().to_string(), Box::new(message)));
+  }
+
+  fn display_all<T: Debug>(iter: Iter<T>) -> String {
```

Comment:
`Vec` (and other stdlib types) implement `Debug` for members which implement `Debug`, so rather than this method you should be able to just debug-format a vec directly.:
```
let v: Vec<T> = ???;
??? = format!("{:?}", v);
```
... unless it's not readable on one line, I suppose.

Reply:
I want to newline-separate the entries, because proto formatting is a little sparse with no real boundaries between entries:

One-line:
```
Expected 2 responses but only issued 2. Expected responses:
[name: "gimme-foo", name: "gimme-foo" done: true response {type_url: "type.googleapis.com/google.devtools.remoteexecution.v1test.ExecuteResponse" value: "\n\005*\003foo"}]
Received requests:
[("ExecuteRequest", action {command_digest {hash: "b0eaf01f93bbe3fa0a5d70489ae88f2cfb144cda494a52af3985f4f1fd8755c8" size_bytes: 20}}), ("GetOperationRequest", name: "gimme-foo")]
```

Multi-line:
```
Expected responses:
name: "gimme-foo"
name: "gimme-foo" done: true response {type_url: "type.googleapis.com/google.devtools.remoteexecution.v1test.ExecuteResponse" value: "\n\005*\003foo"}

Received requests:
("ExecuteRequest", action {command_digest {hash: "b0eaf01f93bbe3fa0a5d70489ae88f2cfb144cda494a52af3985f4f1fd8755c8" size_bytes: 20}})
("GetOperationRequest", name: "gimme-foo")
```